[ROCm] Fix AMD GPU alloca address space errors (#433)

phambinhfin opened the pull request:
* Fix AMD GPU alloca address space errors

AMD GPUs require stack allocations (alloca instructions) to be in address space 5 (private/local memory), not address space 0 (generic memory)

* Optimize AMDGPU allocas by keeping AS5 pointers throughout

* Fix AMDGPU allocas to use address space 5 in MLIR lowering

phambinhfin committed:
Fix AMD GPU alloca address space errors (#433)

* Fix AMD GPU alloca address space errors

AMD GPUs require stack allocations (alloca instructions) to be in
address space 5 (private/local memory), not address space 0 (generic
memory)

* Optimize AMDGPU allocas by keeping AS5 pointers throughout

* Fix AMDGPU allocas to use address space 5 in MLIR lowering

dimitar-asenov reviewed (changes requested):
Please add a test verifying the new behavior.

phambinhfin commented on xla/codegen/emitters/transforms/lower_to_llvm.cc:
```diff
@@ -160,6 +160,29 @@ class LowerToLLVMPass : public impl::LowerToLLVMPassBase<LowerToLLVMPass> {
       signalPassFailure();
     }
+    
+    // For AMDGPU, fix allocas to use address space 5 (private)
```
Could you check again

dimitar-asenov commented on xla/codegen/emitters/transforms/lower_tensors.cc:
```diff
@@ -1405,9 +1405,39 @@ class LowerTensorsPass : public impl::LowerTensorsPassBase<LowerTensorsPass> {
  private:
+  // Ensure AMDGPU allocas use address space 5 (private).
+  // AMDGPU requires allocas in AS5, but MLIR lowering creates them in AS0.
+  void EnsureAMDGPUAllocasUseAS5() {
```
Please don't  duplicate this function. Instead, create two new files, e.g. `lowering_utils.h`/`.cc` and add this functionality there.

Sorry I didn't provide this comment earlier.

dimitar-asenov reviewed (approved):
Thank you for all the fixes.

phambinhfin committed:
Add milr testcase to check address space 5

dimitar-asenov commented:
@phambinhfin thanks for streamlining this. After the PR is approved (I just approved it), we have an internal process that merges it, so you don't have to do anything. The process could take a few days. If it's not merged by mid next week you can ping me here.

thomasjoerg reviewed (changes requested):
This PR breaks many BUILD targets (e.g. //xla/tools:hlo-opt)

```
[xla/codegen/emitters/transforms/lower_to_llvm.cc:68]: error: no member named 'ROCDL' in namespace 'mlir'
  420 |     registry.insert<mlir::ROCDL::ROCDLDialect>();
```

phambinhfin committed:
Add ROCDL/NVVM dialect includes to passes.h for pass registration

phambinhfin committed:
Fix MLIR test for AMD GPU alloca address space

- Fix pass name: xla-lower-to-llvm -> xla-lower-to-llvm-gpu
- Remove invalid target_type option (not supported by LowerToLLVMGPUPass)
- Add ROCDLDialect to dependent dialects and includes
- Register LowerToLLVMGPUPass in emitters_opt tool
